### PR TITLE
OPENEUROPA-1946: [oe_media] Document media type should use the private file system.

### PR DESCRIPTION
**Pull request by sergepavle**
### Description

Ensure that the uploaded files go to the private file system
Ensure that access to these files is controlled by the access to the node(s) that reference the media. Ask reported about implementation

### Change log

- Added:
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: oe_media.module [hunk 1]
```diff
@@ -16,3 +16,98 @@ declare(strict_types = 1);
 function oe_media_media_source_info_alter(array &$sources): void {
   $sources['oembed:video']['providers'] = ['YouTube', 'Vimeo', 'Dailymotion'];
 }
+
+/**
+ * Implements hook_file_download().
+ *
+ * Allow file downloading only if user have view access to parent entity (node).
```
Comment (imanoleguskiza) [inline oe_media.module]: I would reword this to something like: "Only allow file downloads if the user has view access to any parent node entity"

File: oe_media.module [hunk 2]
```diff
@@ -16,3 +16,98 @@ declare(strict_types = 1);
...
+      // same user. This prevents the security issue that a private file that
+      // was protected by field permissions becomes available after its usage
+      // was removed and before it is actually deleted from the file system.
+      // Modules that depend on this behavior should make the file permanent
+      // instead.
+      return -1;
+    }
+  }
+
+  if ($file->isPermanent()) {
```
Comment (imanoleguskiza) [inline oe_media.module]: Just curious to know if this is actually required. If a file is not temporary surely is permanent? Or is there any other possible state?

File: oe_media.module [hunk 3]
```diff
@@ -16,3 +16,98 @@ declare(strict_types = 1);
...
+    $field_referenced_to_media = [];
+    /** @var \Drupal\field\Entity\FieldConfig $field */
+    foreach ($fields as $field) {
+      $field_settings = $field->getSettings();
+      if ($field_settings['handler'] === 'default:media' && in_array('document', $field_settings['handler_settings']['target_bundles'])) {
+        $field_referenced_to_media[] = $field->getName();
+      }
+    }
+
+    // Get all medias which use this file.
```
Comment (imanoleguskiza) [inline oe_media.module]: Comment is not correct, you are getting nodes here.

File: config/install/field.storage.media.oe_media_file.yml [hunk 4]
```diff
@@ -12,7 +12,7 @@ settings:
   target_type: file
   display_field: false
   display_default: false
-  uri_scheme: public
+  uri_scheme: private
```
Comment (upchuk) [inline config/install/field.storage.media.oe_media_file.yml]: Let's also configure the private file system in the runner:
```
file_private_path: 'sites/default/files/private'
```

File: oe_media.module [hunk 5]
```diff
@@ -16,3 +16,98 @@ declare(strict_types = 1);
 function oe_media_media_source_info_alter(array &$sources): void {
   $sources['oembed:video']['providers'] = ['YouTube', 'Vimeo', 'Dailymotion'];
 }
+
+/**
+ * Implements hook_file_download().
+ *
+ * Allow file downloading only if user have view access to parent entity (node).
+ */
+function oe_media_file_download($uri) {
```
Comment (upchuk) [inline oe_media.module]: This entire thing is not needed. We need to do something else. 

What we need to achieve is that the file is accessible if the user has access to the media entity. And we will then ensure that the user has access to the media entity of the media entity is referenced by at least one published node.

By default, file download access is already set to grant access is there is access to the entity that references it. See `file_file_download` and `FileAccessControlHandler`. So all we need to do is have a hook_entity_access to control access to the document media entity based on access to the referencing nodes:

* If it's not being referenced by a node, we don't do anything. We rely on the publish state of the media entity. The entity itself doesn't have to be referenced by a node
* If it is referenced by at least one node, we grant access if any of the nodes is accessible ($node->access('view')).
